##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
  <td>
    <a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png" />See TF Hub models</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with [TensorFlow Hub](https://tfhub.dev) and Keras.

It uses the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [`tf.keras`](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [`tensorflow_hub`](https://www.tensorflow.org/hub), a library for loading trained models from [TFHub](https://tfhub.dev) in a single line of code. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
!pip install tensorflow-hub
!pip install tensorflow-datasets

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.5.0
Eager mode:  True
Hub version:  0.12.0
GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [5]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. Use a pre-trained text embedding as the first layer, which will have three advantages:

*   You don't have to worry about text preprocessing,
*   Benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example you use a **pre-trained text embedding model** from [TensorFlow Hub](https://tfhub.dev) called [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2).

There are many other pre-trained text embeddings from TFHub that can be used in this tutorial:

* [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - trained with the same NNLM architecture on the same data as [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2), but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model.
* [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - the same as [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2), but with additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation.
* [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - a much larger model yielding 512 dimensional embeddings trained with a deep averaging network (DAN) encoder.

And many more! Find more [text embedding models](https://tfhub.dev/s?module-type=text-embedding) on TFHub.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423195 , -0.0119017 ,  0.06337538,  0.06862972, -0.16776837,
        -0.10581174,  0.16865303, -0.04998824, -0.31148055,  0.07910346,
         0.15442263,  0.01488662,  0.03930153,  0.19772711, -0.12215476,
        -0.04120981, -0.2704109 , -0.21922152,  0.26517662, -0.80739075,
         0.25833532, -0.3100421 ,  0.28683215,  0.1943387 , -0.29036492,
         0.03862849, -0.7844411 , -0.0479324 ,  0.4110299 , -0.36388892,
        -0.58034706,  0.30269456,  0.3630897 , -0.15227164, -0.44391504,
         0.19462997,  0.19528408,  0.05666234,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201318 , -0.04418665, -0.08550783,
        -0.55847436, -0.23336391, -0.20782952, -0.03543064, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862679,  0.7753425 , -0.07667089,
        -0.15752277,  0.01872335, -0.08169781, -0.3521876 ,  0.4637341 ,
        -0.08492756,  0.07166859, -0.00670817,  0.12686075, -0.19326553,
 

Let's now build the full model:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 50)                48190600  
_________________________________________________________________
dense (Dense)                (None, 16)                816       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 48,191,433
Trainable params: 48,191,433
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that you are using ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`. For this NNLM model, the `embedding_dimension` is 50.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), you'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when you are exploring regression problems (say, to predict the price of a house), you'll see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 10 epochs in mini-batches of 512 samples. This is 10 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


 1/30 [>.............................] - ETA: 51s - loss: 0.6942 - accuracy: 0.4863

 3/30 [==>...........................] - ETA: 1s - loss: 0.6849 - accuracy: 0.5000 

 5/30 [====>.........................] - ETA: 1s - loss: 0.6808 - accuracy: 0.5031

 7/30 [======>.......................] - ETA: 0s - loss: 0.6785 - accuracy: 0.5008

 9/30 [========>.....................] - ETA: 0s - loss: 0.6769 - accuracy: 0.5087

11/30 [==========>...................] - ETA: 0s - loss: 0.6741 - accuracy: 0.5144

13/30 [============>.................] - ETA: 0s - loss: 0.6701 - accuracy: 0.5165

15/30 [==============>...............] - ETA: 0s - loss: 0.6669 - accuracy: 0.5137

17/30 [================>.............] - ETA: 0s - loss: 0.6646 - accuracy: 0.5173

19/30 [==================>...........] - ETA: 0s - loss: 0.6616 - accuracy: 0.5188

21/30 [====================>.........] - ETA: 0s - loss: 0.6584 - accuracy: 0.5246

23/30 [======================>.......] - ETA: 0s - loss: 0.6559 - accuracy: 0.5275

25/30 [========================>.....] - ETA: 0s - loss: 0.6524 - accuracy: 0.5314

27/30 [==========================>...] - ETA: 0s - loss: 0.6494 - accuracy: 0.5353

29/30 [============================>.] - ETA: 0s - loss: 0.6463 - accuracy: 0.5401

30/30 [==============================] - 4s 69ms/step - loss: 0.6458 - accuracy: 0.5411 - val_loss: 0.5995 - val_accuracy: 0.5900


Epoch 2/10


 1/30 [>.............................] - ETA: 30s - loss: 0.5779 - accuracy: 0.6445

 2/30 [=>............................] - ETA: 1s - loss: 0.5831 - accuracy: 0.6094 

 3/30 [==>...........................] - ETA: 1s - loss: 0.5867 - accuracy: 0.6081

 4/30 [===>..........................] - ETA: 1s - loss: 0.5831 - accuracy: 0.6128

 5/30 [====>.........................] - ETA: 1s - loss: 0.5806 - accuracy: 0.6160

 7/30 [======>.......................] - ETA: 1s - loss: 0.5772 - accuracy: 0.6230

 9/30 [========>.....................] - ETA: 0s - loss: 0.5719 - accuracy: 0.6287

11/30 [==========>...................] - ETA: 0s - loss: 0.5683 - accuracy: 0.6365

13/30 [============>.................] - ETA: 0s - loss: 0.5639 - accuracy: 0.6432

15/30 [==============>...............] - ETA: 0s - loss: 0.5616 - accuracy: 0.6527

17/30 [================>.............] - ETA: 0s - loss: 0.5576 - accuracy: 0.6637

19/30 [==================>...........] - ETA: 0s - loss: 0.5538 - accuracy: 0.6669

21/30 [====================>.........] - ETA: 0s - loss: 0.5519 - accuracy: 0.6692

23/30 [======================>.......] - ETA: 0s - loss: 0.5491 - accuracy: 0.6709

25/30 [========================>.....] - ETA: 0s - loss: 0.5446 - accuracy: 0.6759

27/30 [==========================>...] - ETA: 0s - loss: 0.5410 - accuracy: 0.6821

29/30 [============================>.] - ETA: 0s - loss: 0.5367 - accuracy: 0.6894

30/30 [==============================] - 4s 100ms/step - loss: 0.5359 - accuracy: 0.6899 - val_loss: 0.4956 - val_accuracy: 0.7347


Epoch 3/10


 1/30 [>.............................] - ETA: 20s - loss: 0.4628 - accuracy: 0.7793

 3/30 [==>...........................] - ETA: 1s - loss: 0.4634 - accuracy: 0.7591 

 5/30 [====>.........................] - ETA: 1s - loss: 0.4613 - accuracy: 0.7715

 7/30 [======>.......................] - ETA: 0s - loss: 0.4525 - accuracy: 0.7857

 9/30 [========>.....................] - ETA: 0s - loss: 0.4463 - accuracy: 0.7930

11/30 [==========>...................] - ETA: 0s - loss: 0.4427 - accuracy: 0.7951

13/30 [============>.................] - ETA: 0s - loss: 0.4377 - accuracy: 0.7978

15/30 [==============>...............] - ETA: 0s - loss: 0.4338 - accuracy: 0.7992

17/30 [================>.............] - ETA: 0s - loss: 0.4269 - accuracy: 0.8017

19/30 [==================>...........] - ETA: 0s - loss: 0.4223 - accuracy: 0.8049

21/30 [====================>.........] - ETA: 0s - loss: 0.4187 - accuracy: 0.8090

23/30 [======================>.......] - ETA: 0s - loss: 0.4151 - accuracy: 0.8106

25/30 [========================>.....] - ETA: 0s - loss: 0.4107 - accuracy: 0.8138

27/30 [==========================>...] - ETA: 0s - loss: 0.4085 - accuracy: 0.8155

29/30 [============================>.] - ETA: 0s - loss: 0.4053 - accuracy: 0.8167

30/30 [==============================] - 3s 66ms/step - loss: 0.4049 - accuracy: 0.8167 - val_loss: 0.3975 - val_accuracy: 0.8207


Epoch 4/10


 1/30 [>.............................] - ETA: 25s - loss: 0.3319 - accuracy: 0.8652

 3/30 [==>...........................] - ETA: 1s - loss: 0.3186 - accuracy: 0.8874 

 5/30 [====>.........................] - ETA: 1s - loss: 0.3209 - accuracy: 0.8859

 7/30 [======>.......................] - ETA: 0s - loss: 0.3244 - accuracy: 0.8839

 9/30 [========>.....................] - ETA: 0s - loss: 0.3191 - accuracy: 0.8861

11/30 [==========>...................] - ETA: 0s - loss: 0.3194 - accuracy: 0.8807

13/30 [============>.................] - ETA: 0s - loss: 0.3176 - accuracy: 0.8795

15/30 [==============>...............] - ETA: 0s - loss: 0.3120 - accuracy: 0.8822

17/30 [================>.............] - ETA: 0s - loss: 0.3077 - accuracy: 0.8859

19/30 [==================>...........] - ETA: 0s - loss: 0.3042 - accuracy: 0.8881

21/30 [====================>.........] - ETA: 0s - loss: 0.3018 - accuracy: 0.8866

23/30 [======================>.......] - ETA: 0s - loss: 0.2998 - accuracy: 0.8863

25/30 [========================>.....] - ETA: 0s - loss: 0.2968 - accuracy: 0.8870

27/30 [==========================>...] - ETA: 0s - loss: 0.2951 - accuracy: 0.8866

29/30 [============================>.] - ETA: 0s - loss: 0.2931 - accuracy: 0.8872

30/30 [==============================] - 3s 70ms/step - loss: 0.2929 - accuracy: 0.8873 - val_loss: 0.3465 - val_accuracy: 0.8527


Epoch 5/10


 1/30 [>.............................] - ETA: 20s - loss: 0.2701 - accuracy: 0.9043

 3/30 [==>...........................] - ETA: 1s - loss: 0.2453 - accuracy: 0.9180 

 5/30 [====>.........................] - ETA: 1s - loss: 0.2400 - accuracy: 0.9176

 7/30 [======>.......................] - ETA: 1s - loss: 0.2331 - accuracy: 0.9194

 9/30 [========>.....................] - ETA: 0s - loss: 0.2307 - accuracy: 0.9162

11/30 [==========>...................] - ETA: 0s - loss: 0.2313 - accuracy: 0.9148

13/30 [============>.................] - ETA: 0s - loss: 0.2308 - accuracy: 0.9139

15/30 [==============>...............] - ETA: 0s - loss: 0.2294 - accuracy: 0.9151

17/30 [================>.............] - ETA: 0s - loss: 0.2269 - accuracy: 0.9180

19/30 [==================>...........] - ETA: 0s - loss: 0.2269 - accuracy: 0.9178

21/30 [====================>.........] - ETA: 0s - loss: 0.2244 - accuracy: 0.9191

23/30 [======================>.......] - ETA: 0s - loss: 0.2225 - accuracy: 0.9193

25/30 [========================>.....] - ETA: 0s - loss: 0.2200 - accuracy: 0.9215

27/30 [==========================>...] - ETA: 0s - loss: 0.2175 - accuracy: 0.9229

29/30 [============================>.] - ETA: 0s - loss: 0.2158 - accuracy: 0.9237

30/30 [==============================] - 3s 66ms/step - loss: 0.2158 - accuracy: 0.9233 - val_loss: 0.3199 - val_accuracy: 0.8576


Epoch 6/10


 1/30 [>.............................] - ETA: 32s - loss: 0.1691 - accuracy: 0.9492

 2/30 [=>............................] - ETA: 1s - loss: 0.1628 - accuracy: 0.9531 

 3/30 [==>...........................] - ETA: 1s - loss: 0.1598 - accuracy: 0.9505

 4/30 [===>..........................] - ETA: 1s - loss: 0.1672 - accuracy: 0.9458

 6/30 [=====>........................] - ETA: 1s - loss: 0.1714 - accuracy: 0.9447

 8/30 [=======>......................] - ETA: 1s - loss: 0.1704 - accuracy: 0.9451

10/30 [=========>....................] - ETA: 0s - loss: 0.1671 - accuracy: 0.9469

12/30 [===========>..................] - ETA: 0s - loss: 0.1642 - accuracy: 0.9476

14/30 [=============>................] - ETA: 0s - loss: 0.1626 - accuracy: 0.9482

16/30 [===============>..............] - ETA: 0s - loss: 0.1628 - accuracy: 0.9478

18/30 [=================>............] - ETA: 0s - loss: 0.1627 - accuracy: 0.9468

20/30 [===================>..........] - ETA: 0s - loss: 0.1623 - accuracy: 0.9470

22/30 [=====================>........] - ETA: 0s - loss: 0.1609 - accuracy: 0.9482

24/30 [=======================>......] - ETA: 0s - loss: 0.1614 - accuracy: 0.9488

26/30 [=========================>....] - ETA: 0s - loss: 0.1604 - accuracy: 0.9495

28/30 [===========================>..] - ETA: 0s - loss: 0.1595 - accuracy: 0.9493

30/30 [==============================] - 3s 67ms/step - loss: 0.1602 - accuracy: 0.9487 - val_loss: 0.3097 - val_accuracy: 0.8625


Epoch 7/10


 1/30 [>.............................] - ETA: 22s - loss: 0.1213 - accuracy: 0.9590

 3/30 [==>...........................] - ETA: 1s - loss: 0.1193 - accuracy: 0.9622 

 5/30 [====>.........................] - ETA: 1s - loss: 0.1191 - accuracy: 0.9648

 7/30 [======>.......................] - ETA: 0s - loss: 0.1218 - accuracy: 0.9651

 9/30 [========>.....................] - ETA: 0s - loss: 0.1237 - accuracy: 0.9642

11/30 [==========>...................] - ETA: 0s - loss: 0.1226 - accuracy: 0.9648

13/30 [============>.................] - ETA: 0s - loss: 0.1221 - accuracy: 0.9636

15/30 [==============>...............] - ETA: 0s - loss: 0.1196 - accuracy: 0.9651

17/30 [================>.............] - ETA: 0s - loss: 0.1201 - accuracy: 0.9635

19/30 [==================>...........] - ETA: 0s - loss: 0.1189 - accuracy: 0.9646

21/30 [====================>.........] - ETA: 0s - loss: 0.1186 - accuracy: 0.9648

23/30 [======================>.......] - ETA: 0s - loss: 0.1170 - accuracy: 0.9659

25/30 [========================>.....] - ETA: 0s - loss: 0.1172 - accuracy: 0.9656

27/30 [==========================>...] - ETA: 0s - loss: 0.1171 - accuracy: 0.9659

29/30 [============================>.] - ETA: 0s - loss: 0.1176 - accuracy: 0.9656

30/30 [==============================] - 3s 67ms/step - loss: 0.1176 - accuracy: 0.9656 - val_loss: 0.3083 - val_accuracy: 0.8681


Epoch 8/10


 1/30 [>.............................] - ETA: 23s - loss: 0.0775 - accuracy: 0.9824

 3/30 [==>...........................] - ETA: 1s - loss: 0.0897 - accuracy: 0.9798 

 5/30 [====>.........................] - ETA: 1s - loss: 0.0919 - accuracy: 0.9777

 7/30 [======>.......................] - ETA: 1s - loss: 0.0898 - accuracy: 0.9791

 9/30 [========>.....................] - ETA: 0s - loss: 0.0873 - accuracy: 0.9811

11/30 [==========>...................] - ETA: 0s - loss: 0.0865 - accuracy: 0.9814

13/30 [============>.................] - ETA: 0s - loss: 0.0867 - accuracy: 0.9800

15/30 [==============>...............] - ETA: 0s - loss: 0.0852 - accuracy: 0.9799

17/30 [================>.............] - ETA: 0s - loss: 0.0853 - accuracy: 0.9795

19/30 [==================>...........] - ETA: 0s - loss: 0.0864 - accuracy: 0.9789

21/30 [====================>.........] - ETA: 0s - loss: 0.0852 - accuracy: 0.9796

23/30 [======================>.......] - ETA: 0s - loss: 0.0852 - accuracy: 0.9792

25/30 [========================>.....] - ETA: 0s - loss: 0.0862 - accuracy: 0.9784

27/30 [==========================>...] - ETA: 0s - loss: 0.0861 - accuracy: 0.9783

29/30 [============================>.] - ETA: 0s - loss: 0.0865 - accuracy: 0.9779

30/30 [==============================] - 3s 64ms/step - loss: 0.0864 - accuracy: 0.9779 - val_loss: 0.3140 - val_accuracy: 0.8671


Epoch 9/10


 1/30 [>.............................] - ETA: 23s - loss: 0.0572 - accuracy: 0.9961

 2/30 [=>............................] - ETA: 1s - loss: 0.0602 - accuracy: 0.9883 

 3/30 [==>...........................] - ETA: 1s - loss: 0.0654 - accuracy: 0.9870

 5/30 [====>.........................] - ETA: 1s - loss: 0.0670 - accuracy: 0.9879

 7/30 [======>.......................] - ETA: 1s - loss: 0.0657 - accuracy: 0.9886

 9/30 [========>.....................] - ETA: 0s - loss: 0.0661 - accuracy: 0.9881

11/30 [==========>...................] - ETA: 0s - loss: 0.0663 - accuracy: 0.9879

13/30 [============>.................] - ETA: 0s - loss: 0.0660 - accuracy: 0.9877

15/30 [==============>...............] - ETA: 0s - loss: 0.0655 - accuracy: 0.9872

17/30 [================>.............] - ETA: 0s - loss: 0.0651 - accuracy: 0.9876

19/30 [==================>...........] - ETA: 0s - loss: 0.0641 - accuracy: 0.9881

21/30 [====================>.........] - ETA: 0s - loss: 0.0647 - accuracy: 0.9871

23/30 [======================>.......] - ETA: 0s - loss: 0.0645 - accuracy: 0.9868

25/30 [========================>.....] - ETA: 0s - loss: 0.0635 - accuracy: 0.9870

27/30 [==========================>...] - ETA: 0s - loss: 0.0627 - accuracy: 0.9877

29/30 [============================>.] - ETA: 0s - loss: 0.0627 - accuracy: 0.9875

30/30 [==============================] - 3s 68ms/step - loss: 0.0627 - accuracy: 0.9875 - val_loss: 0.3216 - val_accuracy: 0.8692


Epoch 10/10


 1/30 [>.............................] - ETA: 18s - loss: 0.0431 - accuracy: 0.9922

 3/30 [==>...........................] - ETA: 1s - loss: 0.0459 - accuracy: 0.9928 

 5/30 [====>.........................] - ETA: 0s - loss: 0.0457 - accuracy: 0.9914

 7/30 [======>.......................] - ETA: 0s - loss: 0.0498 - accuracy: 0.9911

 9/30 [========>.....................] - ETA: 0s - loss: 0.0472 - accuracy: 0.9924

11/30 [==========>...................] - ETA: 0s - loss: 0.0463 - accuracy: 0.9929

13/30 [============>.................] - ETA: 0s - loss: 0.0461 - accuracy: 0.9934

15/30 [==============>...............] - ETA: 0s - loss: 0.0458 - accuracy: 0.9935

17/30 [================>.............] - ETA: 0s - loss: 0.0457 - accuracy: 0.9931

19/30 [==================>...........] - ETA: 0s - loss: 0.0466 - accuracy: 0.9925

21/30 [====================>.........] - ETA: 0s - loss: 0.0462 - accuracy: 0.9925

23/30 [======================>.......] - ETA: 0s - loss: 0.0463 - accuracy: 0.9924

25/30 [========================>.....] - ETA: 0s - loss: 0.0456 - accuracy: 0.9926

27/30 [==========================>...] - ETA: 0s - loss: 0.0455 - accuracy: 0.9924

29/30 [============================>.] - ETA: 0s - loss: 0.0454 - accuracy: 0.9923

30/30 [==============================] - 3s 66ms/step - loss: 0.0457 - accuracy: 0.9922 - val_loss: 0.3329 - val_accuracy: 0.8688


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 2s - loss: 0.3597 - accuracy: 0.8541


loss: 0.360
accuracy: 0.854


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

* For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.
* Try out more [text-related tutorials](https://www.tensorflow.org/hub/tutorials#text-related-tutorials) using trained models from TFHub.